src/tree/tree_manager.rs

```diff
@@ -0,0 +1,322 @@
+use std::collections::{BTreeMap, HashMap};
+use std::sync::Arc;
+use std::time::Duration;
+
+use ethers::abi::{AbiDecode, RawLog};
+use ethers::contract::{EthCall, EthEvent};
+use ethers::providers::Middleware;
+use ethers::types::{Filter, Log, Selector, ValueOrArray, H160, H256, U256};
+use futures::stream::FuturesUnordered;
+use futures::StreamExt;
+use tokio::sync::mpsc::Sender;
+use tokio::task::JoinHandle;
+
+use super::block_scanner::BlockScanner;
+use super::identity_tree::{LeafUpdates, Root};
+use super::Hash;
+use crate::abi::{
+    DeleteIdentitiesCall,
+    DeleteIdentitiesWithDeletionProofAndBatchSizeAndPackedDeletionIndicesAndPreRootCall,
+    RegisterIdentitiesCall, RootAddedFilter, TreeChangedFilter,
+};
+use crate::error::{ok, Log as _};
+
+pub const BLOCK_SCANNER_SLEEP_TIME: u64 = 5;
+
+pub trait TreeVersion: Default {
+    type ChannelData;
+
+    fn spawn<M: Middleware + 'static>(
+        tx: Sender<Self::ChannelData>,
+        block_scanner: Arc<BlockScanner<M>>,
+    ) -> JoinHandle<()>;
+
+    fn tree_changed_signature() -> H256;
+}
+
+pub struct TreeManager<M: Middleware, T: TreeVersion> {
+    pub address: H160,
+    pub block_scanner: Arc<BlockScanner<M>>,
+    pub chain_id: u64,
+    #[allow(dead_code)]
```

Comment:
How come you're not using PhantomData?

Reply:
Good catch, I meant to update to use PhantomData. I just pushed a commit with the changes.